## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf
from sklearn.datasets import make_moons

#  Import and read the charity_data.csv.
data_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
data_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# make the moon data
X, y = make_moons(n_samples=1500, noise=0.08, random_state=1)

# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = data_df.drop(columns=["EIN", "NAME"])
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [3]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts for binning
counts = application_df["APPLICATION_TYPE"].value_counts()
counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = counts[counts < 500].index

replaced_application_df = application_df

# Replace in dataframe
for app in application_types_to_replace:
    replaced_application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
replaced_application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [6]:
replaced_application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [7]:
# Look at CLASSIFICATION value counts for binning
class_counts = application_df['CLASSIFICATION'].value_counts()
class_counts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [8]:
# You may find it helpful to look at CLASSIFICATION value counts >1
class_counts[class_counts > 1]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [9]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = class_counts[class_counts < 1000].index

# Replace in dataframe
for cls in classifications_to_replace:
    replaced_application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
replaced_application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [10]:
replaced_application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [11]:
# Convert categorical data to numeric with `pd.get_dummies`
# Using `drop_first=True` to prevent multicolliniearity
# See https://builtin.com/machine-learning/dummy-variable-trap-pandas
numeric_df = pd.get_dummies(replaced_application_df, drop_first=True)
numeric_df

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_T10  \
0           1      5000              1                     1   
1           1    108590              1                     0   
2           1      5000              0                     0   
3           1      6692              1                     0   
4           1    142590              1                     0   
...       ...       ...            ...                   ...   
34294       1      5000              0                     0   
34295       1      5000              0                     0   
34296       1      5000              0                     0   
34297       1      5000              1                     0   
34298       1  36500179              0                     0   

       APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  \
0                         0                    0                    0   
1                         0                    1                    0   
2                         0                    0                    0   
3                         0                    1                    0   
4                         0                    1                    0   
...                     ...                  ...                  ...   
34294                     0                    0                    1   
34295                     0                    0                    1   
34296                     0                    1                    0   
34297                     0                    0                    0   
34298                     0                    1                    0   

       APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  ...  \
0                        0                    0                    0  ...   
1                        0                    0                    0  ...   
2                        1                    0                    0  ...   
3                        0                    0                    0  ...   
4                        0                    0                    0  ...   
...                    ...                  ...                  ...  ...   
34294                    0                    0                    0  ...   
34295                    0                    0                    0  ...   
34296                    0                    0                    0  ...   
34297                    1                    0                    0  ...   
34298                    0                    0                    0  ...   

       ORGANIZATION_Trust  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  \
0                       0                  0                       0   
1                       0                  1                       0   
2                       0                  0                       0   
3                       1                  0                       1   
4                       1                  0                       0   
...                   ...                ...                     ...   
34294                   0                  0                       0   
34295                   0                  0                       0   
34296                   0                  0                       0   
34297                   0                  0                       0   
34298                   0                  0                       0   

       INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                             0                   0                 0   
1                             0                   0                 0   
2                             0                   0                 0   
3                             0                   0                 0   
4                             1                   0                 0   
...                         ...                 ...               ...   
34294                         0                   0         

In [12]:
# Split our preprocessed data into our features and target arrays
X = numeric_df.drop('IS_SUCCESSFUL', axis=1).to_numpy()
y = numeric_df["IS_SUCCESSFUL"].to_numpy()

print(X)
print('\n\n---\n\n')
print(y)

[[       1     5000        1 ...        0        0        0]
 [       1   108590        0 ...        0        0        0]
 [       1     5000        0 ...        0        0        0]
 ...
 [       1     5000        0 ...        0        0        0]
 [       1     5000        0 ...        0        0        0]
 [       1 36500179        0 ...        0        0        0]]


---


[1 1 0 ... 0 1 0]


In [13]:
# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1, stratify=y)

print(X_train)
print('\n\n---\n\n')
print(X_test)

[[     1  37065      0 ...      0      0      0]
 [     1   5000      0 ...      0      0      0]
 [     1   5000      0 ...      0      0      0]
 ...
 [     1   5000      1 ...      0      0      0]
 [     1 460342      0 ...      0      0      0]
 [     1   5000      0 ...      0      0      0]]


---


[[   1 5000    0 ...    0    0    0]
 [   1 5000    0 ...    0    0    0]
 [   1 5000    0 ...    0    0    0]
 ...
 [   1 5000    0 ...    0    0    0]
 [   1 5000    0 ...    0    0    0]
 [   1 5000    0 ...    0    0    0]]


In [14]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

print(X_train_scaled)
print('\n\n---\n\n')
print(X_test_scaled)

[[ 0.01045683 -0.03166368 -0.12512949 ... -0.06515756 -0.07488171
  -0.02896422]
 [ 0.01045683 -0.03199709 -0.12512949 ... -0.06515756 -0.07488171
  -0.02896422]
 [ 0.01045683 -0.03199709 -0.12512949 ... -0.06515756 -0.07488171
  -0.02896422]
 ...
 [ 0.01045683 -0.03199709  7.99172148 ... -0.06515756 -0.07488171
  -0.02896422]
 [ 0.01045683 -0.02726252 -0.12512949 ... -0.06515756 -0.07488171
  -0.02896422]
 [ 0.01045683 -0.03199709 -0.12512949 ... -0.06515756 -0.07488171
  -0.02896422]]


---


[[ 0.01045683 -0.03199709 -0.12512949 ... -0.06515756 -0.07488171
  -0.02896422]
 [ 0.01045683 -0.03199709 -0.12512949 ... -0.06515756 -0.07488171
  -0.02896422]
 [ 0.01045683 -0.03199709 -0.12512949 ... -0.06515756 -0.07488171
  -0.02896422]
 ...
 [ 0.01045683 -0.03199709 -0.12512949 ... -0.06515756 -0.07488171
  -0.02896422]
 [ 0.01045683 -0.03199709 -0.12512949 ... -0.06515756 -0.07488171
  -0.02896422]
 [ 0.01045683 -0.03199709 -0.12512949 ... -0.06515756 -0.07488171
  -0.02896422]]


## Compile, Train and Evaluate the Model

In [15]:
from keras import callbacks
# make a function that will create a single sequential model with hyperparameter options
# options include - choices of activation functions between layers, numbers of neurons on layers, numbers of layers

def makeModel(hp): # hyperparameters that will be passed in from a Hyperband tuner variable

  # Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
  number_input_features = len(X_train_scaled[0])
  # layer_1 = 20
  # layer_2 = 10

  # set up the sequential model
  nn = tf.keras.models.Sequential()

  # set up choices for the activation functions to be used in the hidden layers
  activationList = hp.Choice('activation', ['relu', 'tanh', 'sigmoid'])

  # set up the keras tuner to decide (loop) through a number of attributes in the first layer - input_dim = 2 - loop through 1 - 10 neurons on the second layer
  nn.add(
      tf.keras.layers.Dense(
          input_dim=number_input_features,
          activation= activationList,
          units=hp.Int('first_units', min_value=1, max_value=10)
        )
      )

  # set up the keras tuner to test different numbers of hidden layers as well as neurons on those hidden layers
  for num in range(hp.Int('num_layers', 1, 6)):

    # add the layer (or layers) with a neuron (or neurons) on each layer
    nn.add(
        tf.keras.layers.Dense(
            input_dim=number_input_features,
            activation=activationList,
            units=hp.Int(
                'units_' + str(num),
                         min_value=1,
                         max_value=10
                )
        )
    )


  # Output the model
  nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

  # Compile the model
  nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

   # Stop early when best accuracy found
  cb = tf.keras.callbacks.EarlyStopping(monitor='accuracy', min_delta=0.01, patience=15, mode='max', baseline=0.72)

  # Train the model
  nn.fit(X_train_scaled, y_train, epochs=25, callbacks=[cb])

  # Check the structure of the model
  nn.summary()

  # return model
  return nn



In [16]:
pip install -q keras-tuner

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 176.1/176.1 kB 4.2 MB/s eta 0:00:00


In [17]:
# import keras tuner modules
import keras_tuner as kt


In [18]:
# use the Hyperband function in order to run the optimizer function

tunerModel = kt.Hyperband(
    # pass in the function the contains the attributes
    makeModel,
    # pass in the val_accuracy objective function
    objective='val_accuracy',
    # tell the max_epochs
    max_epochs=25,
    # tell the number of hyperband iterations
    hyperband_iterations=2
)

Epoch 1/25
858/858 [==============================] - 5s 3ms/step - loss: 0.6914 - accuracy: 0.5323
Epoch 2/25
858/858 [==============================] - 3s 3ms/step - loss: 0.6911 - accuracy: 0.5324
Epoch 3/25
858/858 [==============================] - 3s 4ms/step - loss: 0.6911 - accuracy: 0.5324
Epoch 4/25
858/858 [==============================] - 4s 5ms/step - loss: 0.6911 - accuracy: 0.5324
Epoch 5/25
858/858 [==============================] - 3s 4ms/step - loss: 0.6911 - accuracy: 0.5324
Epoch 6/25
858/858 [==============================] - 3s 4ms/step - loss: 0.6911 - accuracy: 0.5324
Epoch 7/25
858/858 [==============================] - 3s 3ms/step - loss: 0.6911 - accuracy: 0.5324
Epoch 8/25
858/858 [==============================] - 3s 4ms/step - loss: 0.6911 - accuracy: 0.5324
Epoch 9/25
858/858 [==============================] - 4s 5ms/step - loss: 0.6911 - accuracy: 0.5324
Epoch 10/25
858/858 [==============================] - 2s 2ms/step - loss: 0.6911 - accuracy: 0.5324

The following model takes an hour to run:

In [19]:
# use the .search() from keras tuner to search for the best combo of hyperparameters
tunerModel.search(X_train_scaled, y_train, epochs=25, validation_data=[X_test_scaled, y_test])

Trial 60 Complete [00h 02m 04s]
val_accuracy: 0.7380466461181641

Best val_accuracy So Far: 0.7399417161941528
Total elapsed time: 01h 04m 22s


In [20]:
tunerModel.get_best_hyperparameters(1)[0].values

{'activation': 'sigmoid',
 'first_units': 6,
 'num_layers': 2,
 'units_0': 7,
 'units_1': 4,
 'units_2': 4,
 'units_3': 10,
 'units_4': 6,
 'units_5': 2,
 'tuner/epochs': 25,
 'tuner/initial_epoch': 9,
 'tuner/bracket': 2,
 'tuner/round': 2,
 'tuner/trial_id': '0043'}

In [21]:
# use get_best_models to extract the hyper parameters and evaluate the model loss and accuracy
bestModel = tunerModel.get_best_models(1)[0]

modelLoss, modelAccuracy = bestModel.evaluate(X_test_scaled, y_test)

Epoch 1/25
858/858 [==============================] - 3s 3ms/step - loss: 0.6685 - accuracy: 0.6014
Epoch 2/25
858/858 [==============================] - 2s 2ms/step - loss: 0.5852 - accuracy: 0.7256
Epoch 3/25
858/858 [==============================] - 2s 2ms/step - loss: 0.5767 - accuracy: 0.7274
Epoch 4/25
858/858 [==============================] - 2s 2ms/step - loss: 0.5755 - accuracy: 0.7281
Epoch 5/25
858/858 [==============================] - 2s 2ms/step - loss: 0.5748 - accuracy: 0.7280
Epoch 6/25
858/858 [==============================] - 2s 2ms/step - loss: 0.5742 - accuracy: 0.7272
Epoch 7/25
858/858 [==============================] - 2s 2ms/step - loss: 0.5739 - accuracy: 0.7281
Epoch 8/25
858/858 [==============================] - 2s 2ms/step - loss: 0.5734 - accuracy: 0.7287
Epoch 9/25
858/858 [==============================] - 2s 2ms/step - loss: 0.5729 - accuracy: 0.7282
Epoch 10/25
858/858 [==============================] - 2s 2ms/step - loss: 0.5727 - accuracy: 0.7285

In [ ]:
print('Loss:', modelLoss)
print('Accuracy:', modelAccuracy)

In [ ]:
# Export our model to HDF5 file
bestModel.save('Optimization1.h5')

<b>Optimization Attempt #1</b>

APPLICATION_TYPE cutoff = 500

CLASSIFICATION cutoff = 1,000

The accuracy percent shows that approximately 74% of the model's predicted values align with the true values in the original dataset.

Additional attempts could be made to try to reach 75% accuracy, but since this is my 3rd attempt, I will do this outside of homework and class time.